diff --git a/Library/Homebrew/brew.sh b/Library/Homebrew/brew.sh
index 98de43e286..1d97d326c0 100644
--- a/Library/Homebrew/brew.sh
+++ b/Library/Homebrew/brew.sh
@@ -55,6 +55,11 @@ git() {
   "$HOMEBREW_LIBRARY/Homebrew/shims/scm/git" "$@"
 }
 
+vbrew_shim() {
+  PATH="$HOMEBREW_LIBRARY/VirtualBrew/shims:$PATH" "$@"
+  return $?
+}
+
 HOMEBREW_VERSION="$(git -C "$HOMEBREW_REPOSITORY" describe --tags --dirty --abbrev=7 2>/dev/null)"
 HOMEBREW_USER_AGENT_VERSION="$HOMEBREW_VERSION"
 if [[ -z "$HOMEBREW_VERSION" ]]
@@ -69,7 +74,7 @@ then
   odie "Cowardly refusing to continue at this prefix: $HOMEBREW_PREFIX"
 fi
 
-HOMEBREW_SYSTEM="$(uname -s)"
+HOMEBREW_SYSTEM="$(vbrew_shim uname -s)"
 case "$HOMEBREW_SYSTEM" in
   Darwin) HOMEBREW_MACOS="1" ;;
   Linux)  HOMEBREW_LINUX="1" ;;
@@ -77,12 +82,12 @@ esac
 
 if [[ -n "$HOMEBREW_MACOS" ]]
 then
-  HOMEBREW_PROCESSOR="$(uname -p)"
+  HOMEBREW_PROCESSOR="$(vbrew_shim uname -p)"
   HOMEBREW_PRODUCT="Homebrew"
   HOMEBREW_SYSTEM="Macintosh"
   # This is i386 even on x86_64 machines
   [[ "$HOMEBREW_PROCESSOR" = "i386" ]] && HOMEBREW_PROCESSOR="Intel"
-  HOMEBREW_MACOS_VERSION="$(/usr/bin/sw_vers -productVersion)"
+  HOMEBREW_MACOS_VERSION="$(vbrew_shim sw_vers -productVersion)"
   HOMEBREW_OS_VERSION="macOS $HOMEBREW_MACOS_VERSION"
   # Don't change this from Mac OS X to match what macOS itself does in Safari on 10.12
   HOMEBREW_OS_USER_AGENT_VERSION="Mac OS X $HOMEBREW_MACOS_VERSION"
@@ -109,10 +114,10 @@ then
     HOMEBREW_CACHE="$HOME/Library/Caches/Homebrew"
   fi
 else
-  HOMEBREW_PROCESSOR="$(uname -m)"
+  HOMEBREW_PROCESSOR="$(vbrew_shim uname -m)"
   HOMEBREW_PRODUCT="${HOMEBREW_SYSTEM}brew"
-  [[ -n "$HOMEBREW_LINUX" ]] && HOMEBREW_OS_VERSION="$(lsb_release -sd 2>/dev/null)"
-  : "${HOMEBREW_OS_VERSION:=$(uname -r)}"
+  [[ -n "$HOMEBREW_LINUX" ]] && HOMEBREW_OS_VERSION="$(vbrew_shim lsb_release -sd 2>/dev/null)"
+  : "${HOMEBREW_OS_VERSION:=$(vbrew_shim uname -r)}"
   HOMEBREW_OS_USER_AGENT_VERSION="$HOMEBREW_OS_VERSION"
 
   if [[ -z "$HOMEBREW_CACHE" ]]
diff --git a/Library/Homebrew/formula.rb b/Library/Homebrew/formula.rb
index 72bf39c565..e2c3f5519c 100644
--- a/Library/Homebrew/formula.rb
+++ b/Library/Homebrew/formula.rb
@@ -15,6 +15,7 @@ require "migrator"
 require "extend/ENV"
 require "language/python"
 require "tab"
+require "pathname"
 
 # A formula provides instructions and metadata for Homebrew to install a piece
 # of software. Every Homebrew formula is a {Formula}.
@@ -1859,6 +1860,14 @@ class Formula
     $stderr.reopen(out)
     out.close
     args.collect!(&:to_s)
+    homebrew_lib = ENV["HOMEBREW_LIBRARY"]
+    shim_shell = Pathname.new(
+      "#{homebrew_lib}/VirtualBrew/shims/sh"
+    )
+    if shim_shell.exist?
+      args.insert 0, cmd
+      cmd = shim_shell.to_s
+    end
     begin
       exec(cmd, *args)
     rescue
